### PR TITLE
Rewrite C wrapper using Clang

### LINKED ISSUE #8: Interface completion
We already support a lot of the interface, so lets push for 100%. This was taken from the 5.6 Reference Manual: http://www.gurobi.com/documentation/5.6/reference-manual/refman
### Environment Creation and Destruction
- [x] GRBloadenv (as the constructor for Env - consider adding a gurobi_env() function which does same for consistency with style of rest of API, esp. model constructor. Should probably support logfilename as optional argument)
- [ ] GRBloadclientenv (compute server-related - ignore)
- [x] GRBfreeenv
- [ ] GRBgetconcurrentenv
- [ ] GRBdiscardconcurrentenvs
### Model Creation and Modification
- [ ] GRBloadmodel
- [x] GRBnewmodel
- [x] GRBcopymodel
- [x] GRBaddconstr
- [x] GRBaddconstrs
- [x] GRBaddqconstr
- [x] GRBaddqpterms
- [x] GRBaddrangeconstr
- [x] GRBaddrangeconstrs
- [ ] GRBaddsos
- [x] GRBaddvar
- [x] GRBaddvars
- [ ] GRBchgcoeffs
- [ ] GRBdelconstrs
- [ ] GRBdelq
- [ ] GRBdelqconstrs
- [ ] GRBdelsos
- [ ] GRBdelvars
- [x] GRBupdatemodel
- [x] GRBfreemodel
## Model Solution
- [x] GRBoptimize
- [x] GRBcomputeIIS
- [ ] GRBfeasrelax
- [ ] GRBfixedmodel
- [ ] GRBresetmodel
- [ ] GRBsync
### Model Queries
- [ ] GRBgetcoeff
- [ ] GRBgetconstrbyname
- [x] GRBgetconstrs
- [ ] GRBgetenv
- [x] GRBgetq
- [ ] GRBgetqconstr
- [ ] GRBgetsos
- [ ] GRBgetvarbyname
- [ ] GRBgetvars
### Input/Output
- [x] GRBreadmodel
- [ ] GRBread
- [ ] GRBwrite
### Attribute Management
- [ ] GRBgetattrinfo
- [ ] GRBgetintattr
- [ ] GRBsetintattr
- [ ] GRBgetintattrelement
- [ ] GRBsetintattrelement
- [ ] GRBgetintattrarray
- [ ] GRBsetintattrarray
- [ ] GRBgetintattrlist
- [ ] GRBsetintattrlist
- [ ] GRBgetdblattr
- [ ] GRBsetdblattr
- [ ] GRBgetdblattrelement
- [ ] GRBsetdblattrelement
- [ ] GRBgetdblattrarray
- [ ] GRBsetdblattrarray
- [ ] GRBgetdblattrlist
- [ ] GRBsetdblattrlist
- [ ] GRBgetcharattrelement
- [ ] GRBsetcharattrelement
- [ ] GRBgetcharattrarray
- [ ] GRBsetcharattrarray
- [ ] GRBgetcharattrlist
- [ ] GRBsetcharattrlist
- [ ] GRBgetstrattr
- [ ] GRBsetstrattr
- [ ] GRBgetstrattrelement
- [ ] GRBsetstrattrelement
- [ ] GRBgetstrattrarray
- [ ] GRBsetstrattrarray
- [ ] GRBgetstrattrlist
- [ ] GRBsetstrattrlist
### Parameter Management and Tuning
- [ ] GRBtunemodel
- [ ] GRBgettuneresult
- [ ] GRBgetdblparam
- [ ] GRBgetintparam
- [ ] GRBgetstrparam
- [ ] GRBsetdblparam
- [ ] GRBsetintparam
- [ ] GRBsetstrparam
- [ ] GRBgetdblparaminfo
- [ ] GRBgetintparaminfo
- [ ] GRBgetstrparaminfo
- [ ] GRBreadparams
- [ ] GRBwriteparams
### Monitoring Progress - Logging and Callbacks
- [ ] GRBmsg
- [x] GRBsetcallbackfunc
- [ ] GRBgetcallbackfunc
- [x] GRBcbget
- [x] GRBversion
### Modifying Solver Behavior - Callbacks
- [x] GRBcbcut
- [x] GRBcblazy
- [x] GRBcbsolution
- [ ] GRBterminate
### Error Handling
- [x] GRBgeterrormsg
### Advanced simplex routines
- [ ] GRBFSolve
- [ ] GRBBSolve
- [ ] GRBinvColj
- [ ] GRBinvRowi
- [ ] GRBgetBasisHead

### LINKED ISSUE #14: handling range constraints in mathprog interface
The gurobi function `addrangeconstrs` introduces an additional slack variable to the problem for each constraint added. To properly support this, we need to transform user-facing variable indices to Gurobi variable indices at the mathprog interface layer (or convince Gurobi to make this transparent at the C-interface level). Until then, I'm printing a appropriate warning when range constraints are used.

### LINKED ISSUE #18: Document solve status codes
JuliaOpt/JuMP.jl#141

### LINKED ISSUE #24: Finalizer for Env
Eli complained today about memory becoming saturated with a (small) model in a tight loop. I thought this was fixed when we changed the behavior in CPLEX.jl, but it doesn't look like it. 

cc @mlubin @IainNZ

### LINKED ISSUE #25: Unable to set Integer options to infinity
Because Inf is floating point, this code in grb_params.jl fails when trying to set something in GRB_INT_PARAMS to infinity.

```
function set_int_param!(env::Env, name::ASCIIString, v::Integer)
ret = @grb_ccall(setintparam, Cint, (Ptr{Void}, Ptr{Cchar}, Cint),
env, name, v)
if ret != 0
throw(GurobiError(env, ret))
end
end
```

### LINKED ISSUE #80: getsolution() returns infeasible solution for an unbounded problem
On this problem,
```
Min y
s.t. x >= 1
loadproblem!(m, [1. 0.], [-Inf, -Inf], [Inf, Inf], [0., 1.], [1.], [Inf], :Min)
```
calling `optimize!` returns `:Unbounded` and then `getsolution` gives `[6.93747e-310,0.0]` which is not feasible since `x = 6.93747e-310 < 1` (I am using Gurobi 6.51).

However, it seems that in Gurobi `getsolution` gives a feasible solution even for unbounded problem since for the following problem
```
Max x
s.t. x >= 1
x unbounded
y <= -1
loadproblem!(m, [1. 0.], [-Inf, -Inf], [Inf, -1], [1., 0.], [1], [Inf], :Max)
```
`optimize!` returns `:Unbounded` and then `getsolution` returns a feasible solution.

See https://github.com/JuliaOpt/MathProgBase.jl/pull/144

### LINKED ISSUE #100: copy(model::Model)  does not copy the solution information
If I create the model and then optimize it, Then I copy it with function `copy(model::Model)`, I cannot get solution or basis from the copy of model but I can get from the original model.

### LINKED ISSUE #102: getobjgap not implemented
Hello

I could not get the gap out of a Gurobi model, I don't see the function implemented in the code.

regards

### LINKED ISSUE #107: Deleting constraint
I sometimes want to remove a constraint to replace it by a different constraint not on the same variable. Here's a toy example that works:

```julia
using JuMP
using Gurobi

m = Model(solver=GurobiSolver(MIPFocus = 3))
x = @variable(m)
y = @variable(m)
con = @constraint(m, x<=1)
@constraint(m, x>=0)
@constraint(m, y>=2)
@objective(m, Min, (y-x)^2)
solve(m)
JuMP.setRHS(con, 100) # effectively deleting constraint
solve(m)
```

Unfortunately, this only works because I can choose a large number for `JuMP.setRHS` where I do not expect the constraint to be binding. Attempting `JuMP.setRHS(con, Inf)` leads to the following error message:

```
ERROR: LoadError: Internal error.
Stacktrace:
 [1] updatemodel!(::Gurobi.GurobiMathProgModel) at /home/vtjeng/.julia/v0.6/Gurobi/src/GurobiSolverInterface.jl:310
 [2] setvartype!(::Gurobi.GurobiMathProgModel, ::Array{Symbol,1}) at /home/vtjeng/.julia/v0.6/Gurobi/src/GurobiSolverInterface.jl:459
 [3] #build#119(::Bool, ::Bool, ::JuMP.ProblemTraits, ::Function, ::JuMP.Model) at /home/vtjeng/.julia/v0.6/JuMP/src/solvers.jl:391
 [4] (::JuMP.#kw##build)(::Array{Any,1}, ::JuMP.#build, ::JuMP.Model) at ./<missing>:0
 [5] #solve#116(::Bool, ::Bool, ::Bool, ::Array{Any,1}, ::Function, ::JuMP.Model) at /home/vtjeng/.julia/v0.6/JuMP/src/solvers.jl:168
 [6] solve(::JuMP.Model) at /home/vtjeng/.julia/v0.6/JuMP/src/solvers.jl:150
 [7] include_from_node1(::String) at ./loading.jl:569
 [8] include(::String) at ./sysimg.jl:14
 [9] process_options(::Base.JLOptions) at ./client.jl:305
 [10] _start() at ./client.jl:371
```

Is there any way I can work around this?

### LINKED ISSUE #113: Gurobi throwing exception for attribute 'NumVars' after adding info callback
We are facing problem with Gurobi callback, which throws  - `ErrorException("optimizing: Gurobi.GurobiError(10017,\"Unable to retrieve attribute 'NumVars'\")"`. We commented [grb_callback.jl:145](https://github.com/JuliaOpt/Gurobi.jl/blob/4c3792d5328c7b025de7d76eadabc5ec47a1a703/src/grb_callbacks.jl#L145) and rebuild the package, which makes it work.
We tried to execute the example, which is also throwing the same error.
Please find below the code.

```
using JuMP
using MathProgBase
using Gurobi

m = Model(solver = GurobiSolver(MIPGap=0.1, Cuts=0, VarBranch=1, TimeLimit=3600, OutputFlag=1))

@variable(m, 0 <= x <= 2, Int)
@variable(m, 0 <= y <= 2, Int)

@objective(m, Max, x + 2y)

@constraint(m, y + x <= 3.5)

type NodeData
    time::Float64  # in seconds since the epoch
    node::Int
    obj::Float64
    bestbound::Float64
end

bbdata = NodeData[]

function infocallback(cb)
    node      = MathProgBase.cbgetexplorednodes(cb)
    obj       = MathProgBase.cbgetobj(cb)
    bestbound = MathProgBase.cbgetbestbound(cb)
    push!(bbdata, NodeData(time(),node,obj,bestbound))
end
addinfocallback(m, infocallback, when = :Intermediate)

solve(m)
```

**Library Used: Julia, JuMP**
JuMP.jl Version: 0.17.1
Gurobi.jl Version: 0.3.3

**Language and Solver** 
Julia Version: 0.5.1
Gurobi Version: 7.0.2

### LINKED ISSUE #155: Support general constraints
Does Gurobi.jl support general constraints such as `min`, `max`, `abs`, `or` ... already?

Btw, does JuMP even support them?

http://www.gurobi.com/documentation/7.5/refman/py_model_addgenconstrmin.html

### LINKED ISSUE #160: Code implementation for retrieving the number of produced usercuts in a Infocallback
Hello, 

I would like to propose an addition to the function "MPBWrapper.jl" to retrieve the number of generated usercuts during the B&B search. This is following the discussion at https://discourse.julialang.org/t/does-anyone-know-how-to-retrieve-the-number-of-generated-cuts-in-the-informational-callback/15370

Note that a Method to retrieve the number of generated cuts with Gurobi already exists in function "grb_callbacks.jl" (https://github.com/JuliaOpt/Gurobi.jl/blob/120d606da08facffb206d73d932dbb0bf7d7001d/src/grb_callbacks.jl#L110) 

Then, the only addition to be made is in the wrapper (https://github.com/JuliaOpt/Gurobi.jl/blob/120d606da08facffb206d73d932dbb0bf7d7001d/src/MPBWrapper.jl#L570) 

```
function cbgetmipcutcnt(d::GurobiCallbackData)
    @assert d.state == :Intermediate
    return cbget_mip_cutcnt(d.cbdata, d.where)
end

function cbgetmipcutcnt(d::GurobiCallbackData,output)
    @assert d.state == :Intermediate
    return cbget_mip_cutcnt(d.cbdata, d.where, output)
end
```

Then one could retrieve the number of generated usercuts in an Infocallback as follows:
```
#Produce an info callback
bbdata = Vector{Float64}[]
function infocallback(cb)
node = MathProgBase.cbgetexplorednodes(cb)
obj = MathProgBase.cbgetobj(cb)
bestbound = MathProgBase.cbgetbestbound(cb)
gap=(abs(bestbound-obj)/abs(obj))*100
**numcuts = Gurobi.cbgetmipcutcnt(cb)**
push!(bbdata, [time(), node, obj, bestbound, gap, numcuts])
end
addinfocallback(m, infocallback, when = :Intermediate)
```

I don't know how to make a pull request. Could anyone help me on that?

### LINKED ISSUE #266: Make it easier to understand which functions from MOI wrapper are gurobi's api
for instance:
```julia
function MOI.add_variable(model::Optimizer)
    # Initialize `VariableInfo` with a dummy `VariableIndex` and a column,
    # because we need `add_item` to tell us what the `VariableIndex` is.
    index = CleverDicts.add_item(
        model.variable_info, VariableInfo(MOI.VariableIndex(0), 0)
    )
    info = _info(model, index)
    # Now, set `.index` and `.column`.
    info.index = index
    info.column = length(model.variable_info)
    add_cvar!(model.inner, 0.0)
    _require_update(model)
    return index
end
```

could be:

```julia
function MOI.add_variable(model::Optimizer)
    # Initialize `VariableInfo` with a dummy `VariableIndex` and a column,
    # because we need `add_item` to tell us what the `VariableIndex` is.
    index = CleverDicts.add_item(
        model.variable_info, VariableInfo(MOI.VariableIndex(0), 0)
    )
    info = _info(model, index)
    # Now, set `.index` and `.column`.
    info.index = index
    info.column = length(model.variable_info)
    Gurobi.add_cvar!(model.inner, 0.0)
    _require_update(model)
    return index
end
```

this could make it easier to compare implementations in different solvers...

### LINKED ISSUE #342: Generating a simple problem that reliably triggers hits :UserLimit
I'm on Julia 1.4 but using an older version of the JuMP and Gurobi packages (I'm planning to upgrade soon, but one of my other packages constrains me to using JuMP v0.18.6.):

```
  [2e9cd046] Gurobi v0.8.1
  [4076af6c] JuMP v0.18.6
```

When solving an MILP via JuMP using the `GurobiSolver`, I've found that `getobjbound` appears to ignore any constant term (perhaps related to [this issue](https://github.com/jump-dev/Gurobi.jl/issues/111)) if the objective is a `JuMP.AffExpr` when my solve hits a `:UserLimit`. I haven't been able to confirm whether this is an issue if a) my solve does not hit a `:UserLimit` or b) for `getobjectivevalue`. (I can get around this issue by adding an additional variable constrained to be equal to the objective and optimizing over the variable instead.) 

I wanted to add a test (although I'm still not sure which package that test should go into) to ensure that this is no longer the case, but I haven't been able to reproduce this with a simple example as I can't figure out how to reliably trigger `:UserLimit` with a small problem. Does anyone have a suggestion for a problem that is simple to understand but for which I can get the `:UserLimit` status?

### LINKED ISSUE #344: Errors for Pkg.add("Gurobi")
I completely followed the directions to download and install gurobi but then then i run Pkg.add("Gurobi") it gives me an error message. Let me know what I should do!! 
```Julia
Stacktrace:
│  [1] error(::String) at ./error.jl:33
│  [2] top-level scope at /Users/giuliaalvarenga/.julia/packages/Gurobi/7YNJV/deps/build.jl:154
│  [3] include(::String) at ./client.jl:457
│  [4] top-level scope at none:5
│ in expression starting at /Users/giuliaalvarenga/.julia/packages/Gurobi/7YNJV/deps/build.jl:152
│ Unable to locate Gurobi installation. Running some common diagnostics.
│ 
│ Gurobi.jl only supports the following versions:
│ 
│  - gurobi90
│  - gurobi81
│  - gurobi80
│  - gurobi75
│  - gurobi70
│ 
│ Did you download and install one of these versions from gurobi.com?
│ 
│ 
│ Looking for a version of Gurobi in your path:
│ 
│ 
│ We couldn't find the `GUROBI_HOME` environment variable, but we
│ found this version of Gurobi on your path. Is this version one of
│ the supported versions listed above? If not, you should edit your
│ `PATH` to point to the correct version.
```

### LINKED ISSUE #345: Change termination status to locally solved
This should be `MOI.LOCALLY_SOLVED`
https://github.com/jump-dev/Gurobi.jl/blob/183a1c5916ff6253193e7242eba48a172ea7365f/src/MOI_wrapper.jl#L1969

### LINKED ISSUE #347: Test failure
I'm getting this with a fresh install on macOS and Linux:
```
Academic license - for non-commercial use only
LazyConstraint: Error During Test at /Users/dpo/.julia/packages/Gurobi/7YNJV/test/MOI/MOI_callbacks.jl:57
  Got exception outside of a @test
  TypeError: in typeassert, expected Tuple{Function,Gurobi.Model}, got a value of type Core.Compiler.UseRef
  Stacktrace:
   [1] gurobi_callback_wrapper(::Ptr{Nothing}, ::Ptr{Nothing}, ::Int32, ::Ptr{Nothing}) at /Users/dpo/.julia/packages/Gurobi/7YNJV/src/grb_callbacks.jl:9
   [2] macro expansion at /Users/dpo/.julia/packages/Gurobi/7YNJV/src/grb_common.jl:56 [inlined]
   [3] optimize at /Users/dpo/.julia/packages/Gurobi/7YNJV/src/grb_solve.jl:5 [inlined]
   [4] optimize!(::Gurobi.Optimizer) at /Users/dpo/.julia/packages/Gurobi/7YNJV/src/MOI_wrapper.jl:1920
   [5] top-level scope at /Users/dpo/.julia/packages/Gurobi/7YNJV/test/MOI/MOI_callbacks.jl:87
   [6] top-level scope at /Users/julia/buildbot/worker/package_macos64/build/usr/share/julia/stdlib/v1.5/Test/src/Test.jl:1115
   [7] top-level scope at /Users/dpo/.julia/packages/Gurobi/7YNJV/test/MOI/MOI_callbacks.jl:58
   [8] top-level scope at /Users/julia/buildbot/worker/package_macos64/build/usr/share/julia/stdlib/v1.5/Test/src/Test.jl:1115
   [9] top-level scope at /Users/dpo/.julia/packages/Gurobi/7YNJV/test/MOI/MOI_callbacks.jl:57
   [10] include(::String) at ./client.jl:457
   [11] macro expansion at /Users/dpo/.julia/packages/Gurobi/7YNJV/test/runtests.jl:26 [inlined]
   [12] macro expansion at /Users/julia/buildbot/worker/package_macos64/build/usr/share/julia/stdlib/v1.5/Test/src/Test.jl:1190 [inlined]
   [13] macro expansion at /Users/dpo/.julia/packages/Gurobi/7YNJV/test/runtests.jl:25 [inlined]
   [14] macro expansion at /Users/julia/buildbot/worker/package_macos64/build/usr/share/julia/stdlib/v1.5/Test/src/Test.jl:1115 [inlined]
   [15] top-level scope at /Users/dpo/.julia/packages/Gurobi/7YNJV/test/runtests.jl:25
   [16] include(::String) at ./client.jl:457
   [17] top-level scope at none:6
   [18] eval(::Module, ::Any) at ./boot.jl:331
   [19] exec_options(::Base.JLOptions) at ./client.jl:272
   [20] _start() at ./client.jl:506
```

### PR DESCRIPTION
**Pull request by odow**
Getting this shipshape before I step back from maintaining it. 

<s>I don't know what the deprecation path is. 

The biggest issue is the callback file, which had some constants `CB_MIPSOL`, and some functions like `cbget_xxx`. Is it worth deprecating those and nothing else?</s>

image

Closes #8
Closes #14
Closes #18
Closes #24 
Closes #25
Closes #54 
Closes #80
Closes #100 
Closes #102
Closes #107 
Closes #113
Closes #121 
Closes #155
Closes #160
Closes #218
Closes #266
Closes #342
Closes #344  
Closes #345
Closes #347

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,565 +1,232 @@
...
+This may not be desirable in certain cases, for example when part of a package's
+test suite uses Gurobi as an optional test dependency, but Gurobi cannot be
+installed on a CI server running the test suite. To support this use case, the
+`GUROBI_JL_SKIP_LIB_CHECK` environment variable may be set (to any value) to
+make *Gurobi.jl* installable (but not usable).
 
 ## Use with JuMP
 
-We highly recommend that you use the *Gurobi.jl* package with higher level packages such as [JuMP.jl](https://github.com/jump-dev/JuMP.jl).
+We highly recommend that you use the *urobi.jl package with higher level
```
Comment (rtwalker) [inline README.md:47]:
```suggestion
We highly recommend that you use the *Gurobi.jl* package with higher level
```

**Comment by odow** [comment]
Same deprecation strategy as CPLEX.

```
The C API of Gurobi.jl has been rewritten to expose the complete C API, and
all old functions have been removed.

For example:

    model = Gurobi.Optimizer()
    stat = Gurobi.get_status_code(model.inner)

is now:

    model = Gurobi.Optimizer()
    valueP = Ref{Cint}()
    ret = GRBgetintattr(model, "Status", valueP)
    if ret != 0
        # Do something because the call failed
    end
    stat = valueP[]

The new API is more verbose, but the names and function arguments are now
identical to the C API, documentation for which is available at:
https://www.gurobi.com/documentation/9.0/refman/c_api_details.html

To revert to the old API, use:

    import Pkg
    Pkg.add(Pkg.PackageSpec(name = \"Gurobi\", version = v\"0.8.1\"))

Then restart Julia for the change to take effect.
```
image

**Comment by odow** [comment]
This PR is a little hard to review, so I propose merging into `master` to make it easier for people to try out, with the view to waiting a while before tagging and releasing. Then any extra changes will be much easier to review.

**Comment by odow** [comment]
Merging with the same plan as CPLEX: https://github.com/jump-dev/CPLEX.jl/pull/316#issuecomment-700323672

**Pull request merged**